### DISCUSSION TITLE
Italian Layout

### DISCUSSION THREAD

**Discussion by flauta** [8VIM/8VIM#255] [Language Layout]
Actually the Italian keyboard disposition adds the accented chars to the standard ones, but does not take account of the different frequency of the letters in italian texts. 

To facilitate a fast typing, if a letter is more frequent IMO it should be obtained with a shorter move: so it should appear in a inner position in the branches. 

The **Y** letter is first (nearer) in the top-left branch, but **it does not appear at all in the Italian alphabet** (except in foreign words of course). On the other side, letter **R** is the 7th most used letter but it appears as the most distant in the branch.
[eg see there](https://www.sttmedia.com/characterfrequency-italian)
[Wikipedia](https://it.wikipedia.org/wiki/Analisi_delle_frequenze)

So I suggest to swap Y and R position in the Italian keyboard map. 
Other possible useful swaps according with letter frequencies in Italian are Q - X (X is another non-Italian letter) and S - L

I have tried these swaps with the "load local map" function and I find them useful.

**Comment by flide** [comment]
Would you be willing to do the research and come up with a better layout
for italian language? we can make the change and would be able to defend
the layout if it's backed by data.

**Comment by flauta** [comment]
The two links I included in my proposal already contain data about letter frequency in Italian language. See there:

https://www.sttmedia.com/characterfrequency-italian
https://www.liquisearch.com/letter_frequency/relative_frequencies_of_letters_in_other_languages

I don't know if you and the other developers agree with the concept that _more frequent letters_ should be _nearer to the central dot_. IMO such a disposition should accelerate typing since the most frequent letters are obtained with shorter movements.

So my proposal for the italian layout is this: 
```
<lowerCase>nomufv!weishkj@,tlcdzg.'rabqpxy?</lowerCase>
<upperCase>NOMUFV!WEISHKJ@_TLCDZG-"RABQPXY*</upperCase>
```

**Comment by davidoskky** [comment]
More frequent letters are important, however also frequent sequences of letters are important.
I remember that @Glitchy-Tozier developed some software to optimize the keyboard layout taking those in consideration as well.
If I'm not mistaken the mechanism of this software was explained in #90 you might take a look at that to see if the generated layout is better.

**Comment by Glitchy-Tozier** [comment]
If you want, I can see what I get when optimizing for the Italian language!

The main problem with my script is that it ignores diacritics. It'll most likely still provide you with way better results compared to the default layout though.

You can find a post regarding my script here:   https://github.com/flide/8VIM/discussions/138#discussion-64700

**Comment by davidoskky** [comment]
Great, diacritics should not matter as those are added always in the same way, thus you simply have to assign the most frequent to the easiest combination.
The problem now would be finding (or generating) the most common Italian bigrams.

**Comment by Glitchy-Tozier** [comment]
Right, this is a problem. [The website I used for other langues](http://www.practicalcryptography.com/cryptanalysis/letter-frequencies-various-languages/) doesn't seem to feature italian bigrams, so I'm not sure where to find them.

**Comment by flauta** [comment]
Here is one italian digrams table (the only I found):

http://www.crittologia.eu/critto/php/frequenze_bigrammi_lingua.phtml

as for the diacritics: the actual solution ("à" beginning on the same side/place of "a" and so on) is fine for me.

**Comment by Glitchy-Tozier** [comment]
Okay, so with some formatting, the list looks like this. Would you guys agree that this is representative of Italian?
(Ordered from most common to least common)
```
ER 0.018158
ON 0.017247
DI 0.016616
EL 0.015821
ES 0.015471
TE 0.015447
EN 0.015400
RE 0.014653
AN 0.014536
AL 0.013835
RA 0.013087
AR 0.012596
NO 0.012526
IN 0.012526
LE 0.012433
CO 0.012152
DE 0.012059
LA 0.011802
NT 0.011708
TA 0.011638
NE 0.011474
ED 0.011358
IO 0.011311
LL 0.011124
RI 0.010937
IA 0.010914
OS 0.010843
TO 0.010656
SI 0.010446
ST 0.010212
OR 0.009932
AT 0.009348
OL 0.009184
IC 0.009114
ME 0.008553
CH 0.008343
TI 0.008320
LI 0.007899
IS 0.007899
MA 0.007852
IL 0.007735
SE 0.007712
QU 0.007689
TR 0.007338
IM 0.007245
OD 0.007104
ID 0.006941
RO 0.006917
AS 0.006917
SS 0.006754
EC 0.006707
AD 0.006543
HE 0.006497
IE 0.006403
ET 0.006286
NI 0.006263
PE 0.006263
PO 0.006240
AC 0.006029
UN 0.005983
LO 0.005983
SO 0.005983
TT 0.005842
EI 0.005796
PI 0.005702
ND 0.005585
EM 0.005445
OP 0.005235
CI 0.005118
IT 0.005095
UE 0.005024
NA 0.005024
DA 0.004954
AM 0.004837
EA 0.004767
IP 0.004651
RT 0.004651
AP 0.004604
OM 0.004510
GI 0.004464
EP 0.004440
MO 0.004323
OC 0.004253
DO 0.004160
PR 0.004160
CA 0.004136
UA 0.004136
SU 0.004090
ZI 0.004066
OA 0.004043
VE 0.003973
NC 0.003809
EE 0.003762
PA 0.003762
LT 0.003739
CE 0.003529
SC 0.003529
OI 0.003505
IU 0.003459
OT 0.003435
EG 0.003412
SA 0.003318
VA 0.003295
MI 0.003225
OG 0.003131
IV 0.003015
TU 0.002968
MP 0.002968
AG 0.002945
AE 0.002945
IR 0.002921
UR 0.002898
VI 0.002874
CC 0.002828
BI 0.002781
OE 0.002711
FI 0.002664
UT 0.002617
AI 0.002594
HI 0.002594
NS 0.002594
LU 0.002477
IG 0.002454
SP 0.002430
AV 0.002430
EV 0.002407
EO 0.002360
RR 0.002360
AZ 0.002290
CU 0.002267
RS 0.002243
GR 0.002197
IF 0.002173
AA 0.002127
NZ 0.002103
UI 0.002080
UO 0.002057
PP 0.001986
DU 0.001940
II 0.001916
ZA 0.001870
GL 0.001870
EQ 0.001823
BB 0.001799
LP 0.001799
IQ 0.001729
OV 0.001729
EB 0.001636
RM 0.001636
UL 0.001612
FE 0.001612
RN 0.001589
LC 0.001589
GG 0.001589
LS 0.001542
NN 0.001542
NG 0.001542
GE 0.001542
FA 0.001542
BE 0.001519
VO 0.001519
FO 0.001496
LM 0.001496
RC 0.001472
US 0.001449
GO 0.001449
PU 0.001402
NU 0.001379
AF 0.001379
EF 0.001379
AB 0.001355
OF 0.001262
AU 0.001262
UP 0.001239
RV 0.001168
IZ 0.001168
UD 0.001145
NF 0.001145
HA 0.001145
AQ 0.001122
OB 0.001098
GN 0.001075
EU 0.001005
IB 0.001005
RD 0.000982
NM 0.000982
RL 0.000958
CR 0.000958
GU 0.000958
FR 0.000935
PL 0.000935
UC 0.000911
NQ 0.000911
UM 0.000888
OQ 0.000888
BR 0.000841
NP 0.000841
RU 0.000841
MM 0.000795
GA 0.000795
OU 0.000771
GH 0.000748
FF 0.000724
FU 0.000724
RG 0.000701
SF 0.000678
OO 0.000654
AO 0.000654
EZ 0.000631
CL 0.000631
ZO 0.000584
NV 0.000561
ZZ 0.000537
BO 0.000537
MB 0.000514
LG 0.000491
LD 0.000491
ZE 0.000491
BA 0.000491
LN 0.000491
TM 0.000491
RF 0.000491
UB 0.000467
RB 0.000444
EH 0.000444
UG 0.000444
RP 0.000421
LF 0.000397
VV 0.000397
RZ 0.000397
LQ 0.000374
MU 0.000351
UZ 0.000327
UV 0.000304
SM 0.000304
BL 0.000280
LV 0.000280
BU 0.000280
DR 0.000280
CQ 0.000257
IH 0.000234
RQ 0.000234
VR 0.000210
UQ 0.000210
AJ 0.000210
NH 0.000187
FL 0.000164
UF 0.000164
AH 0.000164
OH 0.000164
NL 0.000164
TL 0.000140
DD 0.000140
SQ 0.000140
NR 0.000117
OZ 0.000117
JO 0.000117
HO 0.000117
LR 0.000117
VU 0.000093
DL 0.000093
SL 0.000093
LB 0.000093
EX 0.000070
JA 0.000070
YR 0.000070
TS 0.000070
PS 0.000070
XI 0.000070
SV 0.000070
SH 0.000047
NB 0.000047
PG 0.000047
SD 0.000047
CK 0.000047
PY 0.000047
EW 0.000047
PN 0.000047
TC 0.000047
DH 0.000047
HR 0.000047
IJ 0.000047
YS 0.000047
SR 0.000047
GM 0.000047
SB 0.000047
WE 0.000047
KA 0.000047
JP 0.000047
MS 0.000047
HY 0.000023
CD 0.000023
TN 0.000023
ML 0.000023
ZC 0.000023
YN 0.000023
AX 0.000023
GB 0.000023
RH 0.000023
UX 0.000023
EY 0.000023
SG 0.000023
OX 0.000023
TH 0.000023
MC 0.000023
SK 0.000023
CY 0.000023
AY 0.000023
VC 0.000023
XM 0.000023
AW 0.000023
YE 0.000023
GT 0.000023
LZ 0.000023
XT 0.000023
XA 0.000023
DP 0.000023
KI 0.000023
AK 0.000023
JU 0.000023
DT 0.000023
CT 0.000023
KR 0.000023
CF 0.000023
DS 0.000023
WS 0.000023
DF 0.000023
PF 0.000023
YP 0.000023
ZS 0.000023
YO 0.000023
EK 0.000023
KY 0.000023
GV 0.000023
SW 0.000023
LY 0.000023
PD 0.000023
MY 0.000023
TD 0.000023
DG 0.000023
WO 0.000023
XX 0.000023
```

**Comment by flauta** [comment]
[lista_ita_sp.txt](https://github.com/flide/8VIM/files/8102999/lista_ita_sp.txt)
This is another letters and digrams frequency list I calculated many years ago for another project. 
The words and text source was the PAISA repertory http://www.corpusitaliano.it/
As you can see the results are similar.

**Comment by flauta** [comment]
Sorry the N-grams list from PAISA is this
[lista_paisa_ngrams.txt](https://github.com/flide/8VIM/files/8103010/lista_paisa_ngrams.txt)
(the other one was done on a smaller sample)

**Comment by Glitchy-Tozier** [comment]
Okay, so when doing a **really quick** search, this is the result:
Tomorrow I'll improve the parameters and do a more thorough search.
```
Layout: eotlnrai dcsgumpv whqzfybk ---x--j-
eotlnraidcsgumpvwhqzfybk---x--j-

   ⟍  x                🥊 ⟋   
   🥊 ⟍  z            f ⟋  🥊
     q ⟍  g        u ⟋  y  
       s ⟍  l    n ⟋  m      
         t ⟍     ⟋  r       
             ⟍ ⟋           
             ⟋ ⟍           
         o ⟋     ⟍  a       
       c ⟋  e    i ⟍  p     
     h ⟋  d        v ⟍  b 
   🥊 ⟋  w            k ⟍  j 
   ⟋  🥊                🥊 ⟍ 
─────────────────────────────────────────────> Score: 811485.98    ~90.05 %
─────────────────────────────────────────────> Layout-placing: 1
```
(Boxing-glove (🥊) means that it's a special character: `.`, `,`, `!`, `?`, `'`, `@`)

**Comment by flauta** [comment]
Thanks for your work. I'm going to generate a new XML and start immediately to esperiment your new disposition.

I want to put in another consideration: In the italian language the letter Q is **always** followed by U. It is possible to add a 'diacritic' gesture to get this combination in a single move? The same could be for the combination CH (that is nearly frequent as C + vowel).

**Comment by Glitchy-Tozier** [comment]
> Thanks for your work. I'm going to generate a new XML and start immediately to esperiment your new disposition.

If you wait one more hour, I can give you my final result. :)

> I want to put in another consideration: In the italian language the letter Q is always followed by U. It is possible to add a 'diacritic' gesture to get this combination in a single move? The same could be for the combination CH (that is nearly frequent as C + vowel).

I have no idea, unfortunately.

**Comment by flauta** [comment]
I noticed that in your script output the order of the letters in the result isn't the same of the XML config file.
`eotlnrai dcsgumpv whqzfybk ---x--j-` (first round, second round etc)
in the XML must be converted to
`aipvbkj- eodcwh-- tlsgqz-x nrumfy--` (lower right arm, lower left arm etc)

Could you adapt the output of the script, so that the result can be directly inserted in the XML file?

**Comment by Glitchy-Tozier** [comment]
Should/can the spaces be included?

**Comment by flauta** [comment]
It's not important, since the string must be completed with the missing symbol chars. Let's say that your output is a reference.
But no, in the XML file there are no spaces between the groups. And the string must be duplicated for the capital letters. 
See above an example
https://github.com/flide/8VIM/issues/252#issuecomment-1035439896

**Comment by Glitchy-Tozier** [comment]
There you go.

```
                                                The top 5 BEST layouts:


      ⟍  x                ▓ ⟋
      ▓ ⟍  z            f ⟋  ▓
        q ⟍  g        u ⟋  j
          s ⟍  l    n ⟋  m
            t ⟍     ⟋  r
                ⟍ ⟋
                ⟋ ⟍
            o ⟋     ⟍  a
          c ⟋  e    i ⟍  p
        h ⟋  d        v ⟍  b
      ▓ ⟋  y            k ⟍  w
      ⟋  ▓                ▓ ⟍
aipvbkw- eodcyh-- tlsgqz-x nrumfj--
AIPVBKW- EODCYH-- TLSGQZ-X NRUMFJ--
─────────────────────────────────────────────> Layout-placing: 1
─────────────────────────────────────────────> Score: 811496.75    ~90.05 %


      ⟍  x                ▓ ⟋
      ▓ ⟍  z            f ⟋  y
        q ⟍  g        u ⟋  j
          s ⟍  l    n ⟋  m
            t ⟍     ⟋  r
                ⟍ ⟋
                ⟋ ⟍
            o ⟋     ⟍  a
          c ⟋  e    i ⟍  p
        h ⟋  d        v ⟍  b
      ▓ ⟋  w            k ⟍  ▓
      ⟋  ▓                ▓ ⟍
aipvbk-- eodcwh-- tlsgqz-x nrumfj-y
AIPVBK-- EODCWH-- TLSGQZ-X NRUMFJ-Y
─────────────────────────────────────────────> Layout-placing: 2
─────────────────────────────────────────────> Score: 811495.07    ~90.05 %


      ⟍  x                ▓ ⟋
      ▓ ⟍  z            f ⟋  ▓
        q ⟍  g        u ⟋  y
          s ⟍  l    n ⟋  m
            t ⟍     ⟋  r
                ⟍ ⟋          (This is the one I posted 2h ago)
                ⟋ ⟍
            o ⟋     ⟍  a
          c ⟋  e    i ⟍  p
        h ⟋  d        v ⟍  b
      ▓ ⟋  w            k ⟍  j
      ⟋  ▓                ▓ ⟍
aipvbkj- eodcwh-- tlsgqz-x nrumfy--
AIPVBKJ- EODCWH-- TLSGQZ-X NRUMFY--
─────────────────────────────────────────────> Layout-placing: 3
─────────────────────────────────────────────> Score: 811485.98    ~90.05 %


      ⟍  x                ▓ ⟋
      ▓ ⟍  z            f ⟋  ▓
        q ⟍  g        u ⟋  y
          s ⟍  l    n ⟋  m
            t ⟍     ⟋  r
                ⟍ ⟋
                ⟋ ⟍
            o ⟋     ⟍  a
          c ⟋  e    i ⟍  p
        h ⟋  d        v ⟍  b
      j ⟋  w            k ⟍  ▓
      ⟋  ▓                ▓ ⟍
aipvbk-- eodcwh-j tlsgqz-x nrumfy--
AIPVBK-- EODCWH-J TLSGQZ-X NRUMFY--
─────────────────────────────────────────────> Layout-placing: 4
─────────────────────────────────────────────> Score: 811481.53    ~90.05 %


      ⟍  x                k ⟋
      ▓ ⟍  z            f ⟋  ▓
        q ⟍  g        u ⟋  j
          s ⟍  l    n ⟋  m
            t ⟍     ⟋  r
                ⟍ ⟋
                ⟋ ⟍
            o ⟋     ⟍  a
          c ⟋  e    i ⟍  p
        h ⟋  d        v ⟍  b
      ▓ ⟋  w            y ⟍  ▓
      ⟋  ▓                ▓ ⟍
aipvby-- eodcwh-- tlsgqz-x nrumfjk-
AIPVBY-- EODCWH-- TLSGQZ-X NRUMFJK-
─────────────────────────────────────────────> Layout-placing: 5
─────────────────────────────────────────────> Score: 811479.62    ~90.05 %
#######################################################################################################################
#######################################################################################################################
                                                    Custom layouts:

Old / original 8VIM layout:
nomufv-w eilhkj-- tscdzg-- yabrpxq-
NOMUFV-W EILHKJ-- TSCDZG-- YABRPXQ-
───────────────────────────────────> Score: 756014.82    ~83.90 %
```

**Comment by flide** [comment]
I think it will be great if we can extend your script to generate the final
xml in the end.

**Comment by Glitchy-Tozier** [comment]
I don't think this makes sense at the moment.
1. My script doesn't include special characters (`.,!?@'`)
2. There already is a script that does just that: https://github.com/sslater11/8vim_keyboard_layout_file_generator

**Comment by flauta** [comment]
Thanks for your work! I'm working to the new XML file and I will try it ASAP.

**Comment by flauta** [comment]
[it_regular_italian_NEW.xml.zip](https://github.com/flide/8VIM/files/8124886/it_regular_italian_NEW.xml.zip)
This is the new Italian XML file. Besides the remapping of the keys (according to layout #1 https://github.com/flide/8VIM/issues/252#issuecomment-1046235243 ) I deleted the accented letters that don't exist in Italian and reordered the keys to follow strictly the order of the branches.
@flide I propose you to push this to the source code, if you think it's the right time. Please note: I tested it only via the "Load local file" function, since I'm not able to recompile the project.

**Reply by MassiminoilTrace** [reply]
Wonderful, I was thinking about building a layout too and I found your discussion.
I'm testing it on my phone, loading the "external file" like you

**Comment by davidoskky** [comment]
I find this much better than the previous layout to write in Italian.
It also works decently for English.

**Comment by flide** [comment]
https://github.com/flide/8VIM/pull/327

So, for the Italian users, do you guys agree with the changes made in the PR above? Last comment seems dated.
